### PR TITLE
feat: split forest cli and daemon

### PR DESCRIPTION
**Pull request by hanabi1224**
**Summary of changes**
Changes introduced in this pull request:
- [x] split `forest` crate into `cli`, `daemon` and `shared`
- [x] update `Makefile` accordingly
- [x] update `readme.md` accordingly
- [x] update dockerfile setup accordingly
- [x] update CI setup accordingly

**Reference issue to close (if applicable)**

Closes https://github.com/ChainSafe/forest/issues/1901

**Other information and links**

### CODE REVIEW

File: forest/cli/Cargo.toml [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+cid                              = { workspace = true, default-features = false, features = ["std"] }
+directories                      = "4.0.1"
+fil_actor_miner_v8.workspace     = true
+forest_actor_interface.workspace = true
+forest_auth.workspace            = true
+forest_beacon.workspace          = true
+forest_blocks.workspace          = true
+forest_chain.workspace           = true
+forest_chain_sync.workspace      = true
+forest_cli_shared.path           = "../shared"
```
Comment (tyshko5) [inline forest/cli/Cargo.toml:24]: Should we you write it here or would it be better to move it to top-level `Cargo.toml` as in e.g.lines 112-149

**Comment by tyshko5** [comment]
As far as I understand `forest/cli/src/cli/mod.rs` is mostly old `forest/cli/mod.rs`. Could it be rewired to the old file, so that the history is maintained?

**Comment by tyshko5** [comment]
Same question for`forest/shared/src/cli/client.rs` from https://github.com/ChainSafe/forest/pull/1973#issuecomment-1268187808

**Comment by hanabi1224** [comment]
Hey @tyshko5 that's good points, I have managed to rewired `client.rs` but not `mod.rs` cuz we still have 3 new files and because of https://git.wiki.kernel.org/index.php/GitFaq#Why_does_Git_not_.22track.22_renames.3F it does not seem possible to manually wire it, I've added a comment at the header of the new file with the link to the lost commit history

File: forest/cli/Cargo.toml (hunk 1 above)
Comment (hanabi1224) [reply forest/cli/Cargo.toml:24]: fixed

**Review by lemmih — APPROVED** [review]

File: forest/cli/Cargo.toml [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+[dev-dependencies]
+assert_cmd        = "2"
+quickcheck        = "1"
+quickcheck_macros = "1"
+rand.workspace    = true
+
+[features]
+default = []
+testnet = []
+release = []
```
Comment (LesnyRumcajs) [inline forest/cli/Cargo.toml:91]: I'm not sure, do we need those?

File: forest/shared/Cargo.toml [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+features         = ["easy-errors", "macros", "bytes-v05"]
+version          = "0.11"
+
+[dev-dependencies]
+assert_cmd        = "2"
+quickcheck        = "1"
+quickcheck_macros = "1"
+rand.workspace    = true
+
+[features]
```
Comment (LesnyRumcajs) [inline forest/shared/Cargo.toml:78]:
```suggestion
```

**Review by LesnyRumcajs — APPROVED** [review]
I believe we will need to modify some of the testing infrastructure scripts as well, e.g. https://github.com/ChainSafe/forest/blob/main/scripts/daily_snapshot/upload_snapshot.sh#L25. The PR will be need to made [here](https://github.com/ChainSafe/forest-iac) and merged roughly in the same time.

**Comment by LesnyRumcajs** [comment]
Besides adjusting the infrastructure code to the new API, splendid job @hanabi1224!

File: forest/cli/Cargo.toml (hunk 2 above)
Comment (hanabi1224) [reply forest/cli/Cargo.toml:91]: This original build script uses `release` feature in both readme and CI script
```
- name: Cargo Build
        uses: actions-rs/cargo@v1
        with:
          command: build
          args: --release --bin forest --bin forest-cli --features release
```

File: forest/shared/Cargo.toml (hunk 3 above)
Comment (hanabi1224) [reply forest/shared/Cargo.toml:78]: fixed

File: forest/cli/Cargo.toml (hunk 2 above)
Comment (LesnyRumcajs) [reply forest/cli/Cargo.toml:91]: Are those features used anywhere? If not, it'd be nice to get rid of them, perhaps those are some leftovers.

Comment (lemmih) [reply forest/cli/Cargo.toml:91]: They look unused. I've added them to the list in: https://github.com/ChainSafe/forest/issues/2008

Comment (hanabi1224) [reply forest/cli/Cargo.toml:91]: No neither `release` and `testnet` exist in .rs files

Comment (hanabi1224) [reply forest/cli/Cargo.toml]: fixed

Comment (LesnyRumcajs) [reply forest/cli/Cargo.toml]: Thanks!

**Pull request merged**